### DISCUSSION TITLE
Can bloop support package command?

### DISCUSSION THREAD

**Discussion by davidmlw** [scalacenter/bloop#1251] [General]
It is better to have a bloop package command, so that it is easy for bloop to package a project to jar file. Will this feature be supported?

**Comment by jvican** [comment]
Compiling to jar files will be supported as a mere extension of `bloop compile`, where we currently write class files in a classes directory. However, Bloop will not support a way to customize that step, so this task will not be like `package` in Maven or sbt or Gradle where the user can change what there is in there and how the jar is created. The reasoning behind that is that bloop does not aim to cover all of the use cases of a build tool, but instead focus on compiling, running, testing and debugging Scala and Java code.

**Reply by davidmlw** [reply]
I didn't find the right option of `bloop compile`. When will this mere extension be supported? Before that, is there some workaround that generates jar after bloop compile flow?

**Reply by tgodzik** [reply]
We don't support that currently, however `scala-cli` the new runner for Scala does support it. It's based on Bloop, which does the actual compilation: https://scala-cli.virtuslab.org/

**Comment by bo0ts** [comment]
I'm working on a legacy multi-module maven project and would love to see `bloop package` (even in it's barest form). Multi-module projects are currently not well supported by scala-cli.

**Reply by tgodzik** [reply]
We don't plan to work on it currently. I would just use the default maven mechanisms in that case.

**Reply by bo0ts** [reply]
@tgodzik I would prefer that too, but how do we interface the maven package phase with the compile results of `bloop compile`? bloop compilation is miles faster than using the scala-maven-plugin.

**Reply by tgodzik** [reply]
You would have to create a BSP client that would do interface it, but that is a bit complex and would be the same result as just implementing package command. There is also a notion of straight to jar compilation with zinc (might need some work in Bloop as well), which would then produce jar that can be used later.

There is also an option of migrating automatically to mill which should be faster than maven as well https://mill-build.org/mill/migrating/auto-migrating.html

Not sure which option might help you most, but if some work is needed inside Bloop itself, I don't have time currently to look at it.
